pkg/hds/tracker/healthcheck_generator.go

```diff
@@ -0,0 +1,119 @@
+package tracker
+
+import (
+	"context"
+
+	envoy_core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	envoy_endpoint "github.com/envoyproxy/go-control-plane/envoy/config/endpoint/v3"
+	envoy_service_health "github.com/envoyproxy/go-control-plane/envoy/service/health/v3"
+	"github.com/golang/protobuf/ptypes/wrappers"
+	"google.golang.org/protobuf/types/known/durationpb"
+
+	dp_server "github.com/kumahq/kuma/pkg/config/dp-server"
+	"github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
+	"github.com/kumahq/kuma/pkg/core/resources/manager"
+	"github.com/kumahq/kuma/pkg/core/resources/model"
+	"github.com/kumahq/kuma/pkg/core/resources/store"
+	"github.com/kumahq/kuma/pkg/core/xds"
+	"github.com/kumahq/kuma/pkg/hds/cache"
+	util_xds_v3 "github.com/kumahq/kuma/pkg/util/xds/v3"
+	"github.com/kumahq/kuma/pkg/xds/envoy/names"
+)
+
+type SnapshotGenerator struct {
+	config                  *dp_server.HdsConfig
+	readOnlyResourceManager manager.ReadOnlyResourceManager
+}
+
+func NewSnapshotGenerator(readOnlyResourceManager manager.ReadOnlyResourceManager, config *dp_server.HdsConfig) *SnapshotGenerator {
+	return &SnapshotGenerator{
+		readOnlyResourceManager: readOnlyResourceManager,
+		config:                  config,
+	}
+}
+
+func (g *SnapshotGenerator) GenerateSnapshot(node *envoy_core.Node) (util_xds_v3.Snapshot, error) {
+	proxyID, err := xds.ParseProxyIdFromString(node.Id)
+	if err != nil {
+		return nil, err
+	}
+	dpKey := model.ResourceKey{Mesh: proxyID.Mesh, Name: proxyID.Name}
+	dp := mesh.NewDataplaneResource()
+	if err := g.readOnlyResourceManager.Get(context.Background(), dp, store.GetBy(dpKey)); err != nil {
+		return nil, err
+	}
```

Comment:
Ok, so the cool thing we did when we refactored XDS sync is that we went from querying single DP every 1s to only query when hash mesh is changed. Here we are once again introducing query for single DP every 1s.

Can we change the interval of snapshot generator to like 10s? I expect DP healthchecks to be almost never changed after we start it.